Implemented logic gates

VictorSohier opened the pull request:
## Changes:
Closes #25 
- Gates are displayed
- Tint is for testing purposes
- User input may be handled by invisible image buttons

Closes #44 
- All logic gates implemented

## How to test this PR:
Go to challengeMode.js and swap out this.gate to change behavior. Changing the values between 1 and 0 (or true and false if you prefer) for constant gates changes "input" values to the gate.

The image input is such that the images may be pre-scaled. Show scaled may benefit from caching the scaled output for future use.

## Checklist:
- [x] tested locally
- [x] added tests if needed
- [x] linked the associated ticket
- [x] assigned yourself to this PR

(Side note, this means we can make a computer within a computer)

VictorSohier committed:
Added class files for gates

VictorSohier committed:
Merged from sohier/create-cheatsheet for image scaling function

VictorSohier committed:
forgot to "git add ."

VictorSohier committed:
squashed in #50

VictorSohier committed:
Modified gates to work with modules

VictorSohier committed:
Fixed missed merge conflict

VictorSohier committed:
Added flexibility to gate display code

VictorSohier committed:
Fixed missed merge conflicts

VictorSohier committed:
Added gate test in challengeMode.js

VictorSohier committed:
Added missing image

VictorSohier committed:
Fixed punctuation

rodneybarnes commented on src/objects/gate.js:
```diff
@@ -0,0 +1,36 @@
+
+	answer()
+	{
+		if (this.schema === undefined)
```
I'm not able to see where `schema` is being set/should be set - is this a property on all js classes or something?

VictorSohier replied:
I looked up for a way to make abstract classes in javascript and because there is no official keyword to turn a class abstract, I looked it up and that was what I found.

rodneybarnes commented on src/objects/xnorGate.js:
```diff
@@ -0,0 +1,14 @@
+import { XorGate } from "xorGate.js";
```
I tried swapping in the XnorGate here and got a TypeError in the console:

```
index.html:1 Uncaught TypeError: Failed to resolve module specifier "xorGate.js". Relative references must start with either "/", "./", or "../".
```

When I added a './' to the file path here, I then got a similar error for the import statement in `xorGate.js` for `twoInputGate.js`, and when I fixed that one I saw a separate error related to how that class was being imported:

```
Uncaught SyntaxError: The requested module './twoInputGate.js' does not provide an export named 'TwoInputGate'
```
I removed the curly braces around the import in `xorGate.js` and saw an error in the way I was importing `xnorGate.js` in the challenge mode scene.

I'd recommend going through these classes to make sure they all `export default class`, which will make importing them easier (so we don't have to use named imports for some and not others), and then also testing to make sure they're all wired up correctly (i.e. that the file paths are correct).

VictorSohier replied:
I wasn't paying attention to how I was exporting the gates. This will be fixed by removing the curly Brackets.

rodneybarnes commented on src/scenes/challengeMode.js:
```diff
@@ -1,4 +1,30 @@
+    	this.constant1;
+    	this.constant2;
+    	this.gate;
+    	let img = loadImage("assets/textures/gates/No Input Buffer.png");
```
Is it possible to have each gate we might `new` up here load its own image somehow? It would be nice to be able to call `new AndGate` and it can take care of loading its own image without the dev needing to know which png file to use every time.

VictorSohier replied:
Resolved via a factory object

rodneybarnes reviewed (commented):
Tested locally and the gates show up great - I left a comment about testing the import statements for the other gates to make sure they all work and fixing any that don't.

I like how we're using inheritance, however personally I find our object hierarchy very complex and somewhat confusing. On the one hand, it makes a lot of sense to say that the `answer` for a `xnor` gate is the opposite of the answer for a `xor` gate, however should these objects get any more complex I think the coupling will become an issue and trying to debug them will be challenging as well (for example it took me a bit to unravel all the errors I was getting around the import/export issues I commented on).

I think that refactoring towards a simpler hierarchy would make it easier for other devs to work with, though since the complexity here isn't an immediate issue I think it's safe to leave this as-is. Would you be able to provide some documentation on how to use this? For instance, it looks like the `ConstantGate` is used to represent player input?

Because we're using vanilla Javascript here we don't have a nice type system letting us know what parameters like `input1` and `img` are, so we'll have to augment our efforts with more robust documentation.

rodneybarnes commented on index.html:
```diff
@@ -2,24 +2,13 @@
-        <link rel="stylesheet.css"/>
-        <title>gatehopper</title>
+        <link rel="stylesheet" type="text/css" href="stylesheet.css"/>
+        <title>Gatehopper</title>
```
There seems to be a lot of noise in this PR - not sure why these changes are showing up in this diff, as they're already on `main`. Is this something you would be able to investigate and remedy? That would make it easier to review, so we're not hunting through the diff to see what changes are actually being introduced.

VictorSohier replied:
It might be because I started it before the refactor and now we would be merging it in after the refactor. I might merge again once more to clean it up, especially now that I am using meld rather than a text editor to resolve conflicts.

rodneybarnes commented on src/objects/andGate.js:
```diff
@@ -0,0 +1,14 @@
+import TwoInputGate from "./twoInputGate.js";
```
We seem to have a lot of gate objects, would it be worth it to put them in their own folder inside our objects folder?

VictorSohier replied:
Probably a good idea.

VictorSohier committed:
Added gateFactory, fixed import issues, added a proper test

VictorSohier committed:
cleaned up branch

VictorSohier committed:
Merge branch 'main' into Sohier/logic-gates

VictorSohier committed:
reverted incorrect change to cheatsheet

VictorSohier commented:
Fixed some kind of weird overwriting issue with the cheatsheet, be it something I wrote myself or git not working the way I intended it to, it is now fixed.

rodneybarnes commented on src/objects/gateFactory.js:
```diff
@@ -0,0 +1,92 @@
+   		return promises;
+	}
+
+	create(gateType, input1, input2)
```
I like the way you've set this up!

rodneybarnes commented on src/scenes/challengeMode.js:
```diff
@@ -1,15 +1,49 @@
+					)
+				);
+
+    			for (let i = 2; i < 8; i++)
```
I'm assuming this is just for demonstration purposes?

rodneybarnes reviewed (approved):
Nicely done - looks like all of the gates show up. There are a couple points in the ticket's acceptance criteria I think we'll need to follow up on still, namely that we need a way for the player to input their answer for each logic gate, but I can create a new issue to cover that.

![image](https://user-images.githubusercontent.com/13264219/132136326-9434f3b0-b9b4-4cdc-8827-546adfc901de.png)

VictorSohier committed:
Merge branch 'main' into Sohier/logic-gates
